src/cleaning/acp.c

```diff
@@ -317,18 +317,197 @@ int cleaning_policy_acp_initialize(struct ocf_cache *cache,
+	if (result)
+		return result;
+
 	_acp_rebuild(cache);
 	ocf_kick_cleaner(cache);
 
 	return 0;
 }
 
+#define OCF_ACP_RECOVERY_SHARDS_CNT 32
+
+struct ocf_acp_recovery_context {
+	ocf_cache_t cache;
+
+	struct {
+		uint16_t *chunk[OCF_CORE_MAX];
+		struct {
+			struct list_head chunk_list;
+		} bucket[ACP_MAX_BUCKETS];
+	} shard[OCF_ACP_RECOVERY_SHARDS_CNT];
+
+	ocf_cleaning_recovery_end_t cmpl;
+	void *priv;
+};
+
+static int ocf_acp_recovery_handle(ocf_parallelize_t parallelize,
+		void *priv, unsigned shard_id, unsigned shards_cnt)
+{
+	struct ocf_acp_recovery_context *context = priv;
+	ocf_cache_t cache = context->cache;
+	ocf_cache_line_t entries = cache->device->collision_table_entries;
+	ocf_cache_line_t cline, begin, end, portion;
+	struct acp_cleaning_policy_meta *acp_meta;
+	struct acp_chunk_info *chunk;
+	ocf_core_id_t core_id;
+	uint32_t step = 0;
+
+	portion = DIV_ROUND_UP(entries, shards_cnt);
+	begin = portion*shard_id;
+	end = OCF_MIN(portion*(shard_id + 1), entries);
+
+	for (cline = begin; cline < end; cline++) {
+		ocf_metadata_get_core_and_part_id(cache, cline, &core_id, NULL);
+
+		OCF_COND_RESCHED_DEFAULT(step);
+
+		if (core_id == OCF_CORE_MAX)
+			continue;
+
+		if (!metadata_test_dirty(cache, cline)) {
+			cleaning_policy_acp_init_cache_block(cache, cline);
+			continue;
+		}
+
+		acp_meta = _acp_meta_get(cache, cline);
+		acp_meta->dirty = 1;
+
+		chunk = _acp_get_chunk(cache, cline);
+		context->shard[shard_id].chunk[core_id][chunk->chunk_id]++;
+	}
+
+	return 0;
+}
+
+static void ocf_acp_recovery_chunk(struct ocf_acp_recovery_context *context,
+		struct acp_chunk_info *chunk)
+{
+	ocf_cache_t cache = context->cache;
+	struct acp_context *acp = _acp_get_ctx_from_cache(cache);
+	struct acp_bucket *bucket;
+	unsigned shard_id;
+	uint8_t bucket_id;
+
+	chunk->num_dirty = 0;
+	for (shard_id = 0; shard_id < OCF_ACP_RECOVERY_SHARDS_CNT; shard_id++) {
+		chunk->num_dirty += context->shard[shard_id]
+				.chunk[chunk->core_id][chunk->chunk_id];
+	}
+
+	for (bucket_id = 0; bucket_id < ACP_MAX_BUCKETS; bucket_id++) {
+		bucket = &acp->bucket_info[bucket_id];
+		if (chunk->num_dirty < bucket->threshold)
+			break;
+	}
+
+	bucket = &acp->bucket_info[--bucket_id];
+	chunk->bucket_id = bucket_id;
+
+	list_move_tail(&chunk->list, &bucket->chunk_list);
+}
+
+static void ocf_acp_recovery_finish(ocf_parallelize_t parallelize,
+		void *priv, int error)
+{
+	struct ocf_acp_recovery_context *context = priv;
+	ocf_cache_t cache = context->cache;
+	struct acp_context *acp = _acp_get_ctx_from_cache(cache);
+	ocf_core_id_t core_id;
+	ocf_core_t core;
+	uint64_t core_size;
+	uint64_t num_chunks;
+	uint64_t chunk_id;
+
+	for_each_core(cache, core, core_id) {
+		core_size = core->conf_meta->length;
+		num_chunks = OCF_DIV_ROUND_UP(core_size, ACP_CHUNK_SIZE);
+
+		for (chunk_id = 0; chunk_id < num_chunks; chunk_id++) {
+			ocf_acp_recovery_chunk(context,
+					&acp->chunk_info[core_id][chunk_id]);
+		}
+	}
```

Comment:
How about adding a reschedule? I guess it might take a long time for huge cores.

Reply:
I guess it never hurts. However chunk size is currently 100MiB, so to reach default step threshold (1000000) you'd need to have 100TiB core. It's not technically impossible, but I'd consider it future-proofing at this point :)